io.zeebe.broker.workflow.message.MessageMappingTest flaky

```java
[2019-04-25T12:57:07.877Z] [ERROR] shouldMergeMessageVariablesByDefault[non-interrupting boundary event](io.zeebe.broker.workflow.message.MessageMappingTest)  Time elapsed: 5.071 s  <<< ERROR!

[2019-04-25T12:57:07.877Z] io.zeebe.test.util.stream.StreamWrapperException: No event found matching the criteria

[2019-04-25T12:57:07.877Z] 	at java.util.Optional.orElseThrow(Optional.java:290)

[2019-04-25T12:57:07.877Z] 	at io.zeebe.test.util.stream.StreamWrapper.getFirst(StreamWrapper.java:80)

[2019-04-25T12:57:07.877Z] 	at io.zeebe.broker.workflow.message.MessageMappingTest.shouldMergeMessageVariablesByDefault(MessageMappingTest.java:160)

[2019-04-25T12:57:07.877Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

[2019-04-25T12:57:07.877Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

[2019-04-25T12:57:07.877Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

[2019-04-25T12:57:07.877Z] 	at java.lang.reflect.Method.invoke(Method.java:498)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)

[2019-04-25T12:57:07.877Z] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)

[2019-04-25T12:57:07.877Z] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)

[2019-04-25T12:57:07.877Z] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)

[2019-04-25T12:57:07.877Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)

[2019-04-25T12:57:07.877Z] 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.Suite.runChild(Suite.java:128)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.Suite.runChild(Suite.java:27)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)

[2019-04-25T12:57:07.877Z] 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)

[2019-04-25T12:57:07.877Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)

[2019-04-25T12:57:07.877Z] 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)

[2019-04-25T12:57:07.877Z] 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.Suite.runChild(Suite.java:128)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.Suite.runChild(Suite.java:27)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)

[2019-04-25T12:57:07.877Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)

[2019-04-25T12:57:07.877Z] 	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)

[2019-04-25T12:57:07.936Z] 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

and locally after 209 runs


```java

java.util.concurrent.ExecutionException: Request timed out after PT5S

	at io.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:155)
	at io.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:122)
	at io.zeebe.util.sched.FutureUtil.join(FutureUtil.java:28)
	at io.zeebe.util.sched.future.CompletableActorFuture.join(CompletableActorFuture.java:227)
	at io.zeebe.test.broker.protocol.clientapi.ExecuteCommandRequest.await(ExecuteCommandRequest.java:108)
	at io.zeebe.test.broker.protocol.clientapi.ExecuteCommandRequestBuilder.sendAndAwait(ExecuteCommandRequestBuilder.java:35)
	at io.zeebe.test.broker.protocol.clientapi.ClientApiRule.deployWorkflow(ClientApiRule.java:255)
	at io.zeebe.broker.workflow.message.MessageMappingTest.deployWorkflowWithMapping(MessageMappingTest.java:237)
	at io.zeebe.broker.workflow.message.MessageMappingTest.shouldMergeMessageVariablesByDefault(MessageMappingTest.java:141)
	at sun.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:67)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: io.zeebe.transport.RequestTimeoutException: Request timed out after PT5S
	at io.zeebe.transport.impl.sender.OutgoingRequest.timeout(OutgoingRequest.java:151)
	at io.zeebe.transport.impl.sender.Sender.onTimerExpiry(Sender.java:483)
	at org.agrona.DeadlineTimerWheel.poll(DeadlineTimerWheel.java:284)
	at io.zeebe.transport.impl.sender.Sender.processTimeouts(Sender.java:114)
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90)
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53)
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189)
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154)
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135)
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112)



```

--- a/bpmn-model/src/main/java/io/zeebe/model/bpmn/builder/ProcessBuilder.java
+++ b/bpmn-model/src/main/java/io/zeebe/model/bpmn/builder/ProcessBuilder.java
@@ -23,6 +23,7 @@ import io.zeebe.model.bpmn.instance.SubProcess;
 import io.zeebe.model.bpmn.instance.bpmndi.BpmnShape;
 import io.zeebe.model.bpmn.instance.dc.Bounds;
 import java.util.Collection;
+import java.util.function.Consumer;
 
 /** @author Sebastian Menski */
 public class ProcessBuilder extends AbstractProcessBuilder<ProcessBuilder> {
@@ -60,6 +61,13 @@ public class ProcessBuilder extends AbstractProcessBuilder<ProcessBuilder> {
     resizeSubProcess(targetBpmnShape);
 
     return new EventSubProcessBuilder(modelInstance, subProcess);
+  }
+
+  public ProcessBuilder eventSubProcess(
+      final String id, final Consumer<EventSubProcessBuilder> consumer) {
+    final EventSubProcessBuilder builder = eventSubProcess(id);
+    consumer.accept(builder);
+    return this;
   }
 
   @Override

--- a/engine/src/test/java/io/zeebe/engine/processor/workflow/message/MessageMappingTest.java
+++ b/engine/src/test/java/io/zeebe/engine/processor/workflow/message/MessageMappingTest.java
@@ -12,7 +12,6 @@ import static io.zeebe.test.util.MsgPackUtil.asMsgPack;
 import io.zeebe.engine.util.EngineRule;
 import io.zeebe.model.bpmn.Bpmn;
 import io.zeebe.model.bpmn.BpmnModelInstance;
-import io.zeebe.model.bpmn.builder.ProcessBuilder;
 import io.zeebe.model.bpmn.builder.ZeebeVariablesMappingBuilder;
 import io.zeebe.model.bpmn.instance.BoundaryEvent;
 import io.zeebe.model.bpmn.instance.IntermediateCatchEvent;
@@ -44,18 +43,21 @@ public class MessageMappingTest {
   private static final String PROCESS_ID = "process";
   private static final String MESSAGE_NAME = "message";
   private static final String CORRELATION_VARIABLE = "key";
+
   private static final BpmnModelInstance CATCH_EVENT_WORKFLOW =
       Bpmn.createExecutableProcess(PROCESS_ID)
           .startEvent()
           .intermediateCatchEvent("catch")
           .message(m -> m.name(MESSAGE_NAME).zeebeCorrelationKey(CORRELATION_VARIABLE))
           .done();
+
   private static final BpmnModelInstance RECEIVE_TASK_WORKFLOW =
       Bpmn.createExecutableProcess(PROCESS_ID)
           .startEvent()
           .receiveTask("catch")
           .message(m -> m.name(MESSAGE_NAME).zeebeCorrelationKey(CORRELATION_VARIABLE))
           .done();
+
   private static final BpmnModelInstance INTERRUPTING_BOUNDARY_EVENT_WORKFLOW =
       Bpmn.createExecutableProcess(PROCESS_ID)
           .startEvent()
@@ -64,6 +66,7 @@ public class MessageMappingTest {
           .message(m -> m.name(MESSAGE_NAME).zeebeCorrelationKey(CORRELATION_VARIABLE))
           .endEvent()
           .done();
+
   private static final BpmnModelInstance NON_INTERRUPTING_BOUNDARY_EVENT_WORKFLOW =
       Bpmn.createExecutableProcess(PROCESS_ID)
           .startEvent()
@@ -72,6 +75,7 @@ public class MessageMappingTest {
           .message(m -> m.name(MESSAGE_NAME).zeebeCorrelationKey(CORRELATION_VARIABLE))
           .endEvent()
           .done();
+
   private static final BpmnModelInstance EVENT_BASED_GATEWAY_WORKFLOW =
       Bpmn.createExecutableProcess(PROCESS_ID)
           .startEvent("start")
@@ -88,6 +92,36 @@ public class MessageMappingTest {
           .endEvent("end2")
           .done();
 
+  private static final BpmnModelInstance INTERRUPTING_EVENT_SUBPROCESS_WORKFLOW =
+      Bpmn.createExecutableProcess(PROCESS_ID)
+          .eventSubProcess(
+              "catch",
+              eventSubProcess ->
+                  eventSubProcess
+                      .startEvent()
+                      .message(m -> m.name(MESSAGE_NAME).zeebeCorrelationKey(CORRELATION_VARIABLE))
+                      .interrupting(true)
+                      .endEvent())
+          .startEvent()
+          .serviceTask("task", t -> t.zeebeTaskType("type"))
+          .endEvent()
+          .done();
+
+  private static final BpmnModelInstance NON_INTERRUPTING_EVENT_SUBPROCESS_WORKFLOW =
+      Bpmn.createExecutableProcess(PROCESS_ID)
+          .eventSubProcess(
+              "catch",
+              eventSubProcess ->
+                  eventSubProcess
+                      .startEvent()
+                      .message(m -> m.name(MESSAGE_NAME).zeebeCorrelationKey(CORRELATION_VARIABLE))
+                      .interrupting(false)
+                      .endEvent())
+          .startEvent()
+          .serviceTask("task", t -> t.zeebeTaskType("type"))
+          .endEvent()
+          .done();
+
   @Rule
   public final RecordingExporterTestWatcher recordingExporterTestWatcher =
       new RecordingExporterTestWatcher();
@@ -100,22 +134,6 @@ public class MessageMappingTest {
 
   private String correlationKey;
 
-  private static BpmnModelInstance createEventSubprocessModel() {
-    final ProcessBuilder builder = Bpmn.createExecutableProcess(PROCESS_ID);
-    builder
-        .eventSubProcess("catch")
-        .startEvent("event_sub_start_msg")
-        .interrupting(true)
-        .message(b -> b.name(MESSAGE_NAME).zeebeCorrelationKey(CORRELATION_VARIABLE))
-        .endEvent("event_sub_end_msg");
-
-    return builder
-        .startEvent("start_proc")
-        .serviceTask("task", t -> t.zeebeTaskType("type"))
-        .endEvent("end_proc")
-        .done();
-  }
-
   @Parameters(name = "{0}")
   public static Object[][] parameters() {
     return new Object[][] {
@@ -124,13 +142,22 @@ public class MessageMappingTest {
       {"event-based gateway", EVENT_BASED_GATEWAY_WORKFLOW},
       {"interrupting boundary event", INTERRUPTING_BOUNDARY_EVENT_WORKFLOW},
       {"non-interrupting boundary event", NON_INTERRUPTING_BOUNDARY_EVENT_WORKFLOW},
-      {"interrupting event subprocess", createEventSubprocessModel()}
+      {"interrupting event subprocess", INTERRUPTING_EVENT_SUBPROCESS_WORKFLOW},
+      // TODO (saig0): enable test case when fixing #3552
+      // {"non-interrupting event subprocess", NON_INTERRUPTING_EVENT_SUBPROCESS_WORKFLOW}
     };
   }
 
   @Before
   public void setUp() {
     correlationKey = UUID.randomUUID().toString();
+
+    ENGINE_RULE
+        .message()
+        .withCorrelationKey(correlationKey)
+        .withName(MESSAGE_NAME)
+        .withVariables(asMsgPack("foo", "bar"))
+        .publish();
   }
 
   @Test
@@ -138,20 +165,13 @@ public class MessageMappingTest {
     // given
     deployWorkflowWithMapping(e -> {});
 
+    // when
     final long workflowInstanceKey =
         ENGINE_RULE
             .workflowInstance()
             .ofBpmnProcessId(PROCESS_ID)
             .withVariable(CORRELATION_VARIABLE, correlationKey)
             .create();
-
-    // when
-    ENGINE_RULE
-        .message()
-        .withCorrelationKey(correlationKey)
-        .withName(MESSAGE_NAME)
-        .withVariables(asMsgPack("foo", "bar"))
-        .publish();
 
     // then
     final Record<VariableRecordValue> variableEvent =
@@ -170,20 +190,13 @@ public class MessageMappingTest {
     // given
     deployWorkflowWithMapping(e -> {});
 
+    // when
     final long workflowInstanceKey =
         ENGINE_RULE
             .workflowInstance()
             .ofBpmnProcessId(PROCESS_ID)
             .withVariable(CORRELATION_VARIABLE, correlationKey)
             .create();
-
-    // when
-    ENGINE_RULE
-        .message()
-        .withCorrelationKey(correlationKey)
-        .withName(MESSAGE_NAME)
-        .withVariables(asMsgPack("foo", "bar"))
-        .publish();
 
     // then
     final Record<VariableRecordValue> variableEvent =
@@ -201,20 +214,14 @@ public class MessageMappingTest {
   public void shouldMapMessageVariablesIntoInstanceVariables() {
     // given
     deployWorkflowWithMapping(e -> e.zeebeOutput("foo", MESSAGE_NAME));
+
+    // when
     final long workflowInstanceKey =
         ENGINE_RULE
             .workflowInstance()
             .ofBpmnProcessId(PROCESS_ID)
             .withVariable(CORRELATION_VARIABLE, correlationKey)
             .create();
-
-    // when
-    ENGINE_RULE
-        .message()
-        .withCorrelationKey(correlationKey)
-        .withName(MESSAGE_NAME)
-        .withVariables(asMsgPack("foo", "bar"))
-        .publish();
 
     // then
     final Record<VariableRecordValue> variableEvent =

